mitm.go

```diff
@@ -173,6 +179,57 @@ func (p *Proxy) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	p.proxyMITM(sc, cc)
 }
 
+func isWebSocket(req *http.Request) bool {
+	return strings.Contains(strings.ToLower(req.Header.Get("Upgrade")), "websocket") &&
+		strings.Contains(strings.ToLower(req.Header.Get("Connection")), "upgrade")
+}
+
+// forwardWebSocket forwards a WebSocket connection directly to the
+// source, skipping the request wrapper. Code shamelessly stolen from
+// https://groups.google.com/forum/#!topic/golang-nuts/KBx9pDlvFOc
+func forwardWebSocket(w http.ResponseWriter, req *http.Request) {
+	host, port, err := net.SplitHostPort(req.URL.Host)
+	if err != nil {
+		// Assume there is no port and use default
+		host = req.URL.Host
+		if req.URL.Scheme == "https" {
+			port = "443"
+		} else {
+			port = "80"
+		}
+	}
+	address := net.JoinHostPort(host, port)
+	d, err := net.Dial("tcp", address)
+	if err != nil {
+		log.Printf("forwardWebSocket: error dialing websocket backend %s: %v", address, err)
+		http.Error(w, "No Upstream", 503)
+		return
+	}
+	nc, _, err := w.(http.Hijacker).Hijack()
+	if err != nil {
+		log.Printf("forwardWebSocket: hijack error: %v", err)
+		http.Error(w, "No Upstream", 503)
+		return
+	}
+	defer nc.Close()
+	defer d.Close()
```

Comment:
Can you please get the defers closer to the variable instantiation. If something goes terribly wrong on Hijack(), the dial wouldn't get closed.